common4j/src/main/com/microsoft/identity/common/java/telemetry/adapter/BrokerTelemetryAdapter.java

```diff
@@ -21,16 +21,93 @@
+import com.microsoft.identity.common.java.util.StringUtil;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 import lombok.NonNull;
 
 /**
  * Adapter to aggregate events for broker telemetry. We make use of the {@link TelemetryAggregationAdapter} which
  * aggregates the events into MATS format.
+ * <p>
+ * We then send out the error events as single events since the {@link TelemetryAggregationAdapter} will pick the last error event
+ * and add combine it with the aggregated event. This means that the event sent by {@link TelemetryAggregationAdapter} only contains
+ * the last error that was dispatched.
  */
 public class BrokerTelemetryAdapter extends TelemetryAggregationAdapter {
     public BrokerTelemetryAdapter(@NonNull IBrokerTelemetryObserver observer) {
         super(observer);
     }
+
+    @Override
+    public void process(@NonNull List<Map<String, String>> rawData) {
+        final Map<String, String> aggregatedMap = aggregateEvent(rawData);
+
+        // Filter out the error events first
+        final List<Map<String, String>> errorEvents = filterErrorEvents(rawData, aggregatedMap);
+
+        // count the number of hits of the errors
+        final List<Map<String, String>> countedErrors = countErrors(errorEvents);
+        // publish the event
+        for (Map<String, String> errorEvent : countedErrors) {
+            getObserver().onReceived(errorEvent);
+        }
+
+        // send out the aggregated event separately. This contains the tag of the error event that was
+        // fired last, hence in case of a failure, we know which error caused it.
+        getObserver().onReceived(aggregatedMap);
+    }
+
+    /**
+     * Filters out error events from the list of events
+     */
+    private List<Map<String, String>> filterErrorEvents(List<Map<String, String>> rawData, Map<String, String> aggregatedMap) {
+        final List<Map<String, String>> errorEvents = new ArrayList<>();
+
+        for (Map<String, String> event : rawData) {
+            if (TelemetryEventStrings.EventType.ERROR_EVENT.equals(event.get(TelemetryEventStrings.Key.EVENT_TYPE))) {
+                final Map<String, String> errorEvent = new HashMap<>(aggregatedMap);
+                errorEvent.putAll(event);
+                errorEvent.put(TelemetryEventStrings.Key.CORRELATION_ID, aggregatedMap.get(TelemetryEventStrings.Key.CORRELATION_ID));
+                errorEvent.put(TelemetryEventStrings.Key.IS_ERROR_EVENT, TelemetryEventStrings.Value.TRUE);
+
+                errorEvents.add(errorEvent);
+            }
+        }
+        return errorEvents;
+    }
+
+    /**
+     * Errors with the same tag are essentially the same error, we don't need to send them to telemetry as different errors. Therefore
+     * we shall add an {@link TelemetryEventStrings.Key#ERROR_COUNT} key that denotes the number of hits of the error.
```

Comment:
Ok, I think I see now what we're trying to do here...

Can you add javadoc to the other method to link here?

Reply:
This could happen from utility functions. E.g. we have a function that checks for internet connectivity. 
We could check for internet connectivity multiple times in the acquire token flow. 

If an exception occurs inside this method, it means we shall have an error emitted for every single time this method is invoked.

Reply:
Hmm..I'm not sure how that happens. I mean the first time network check fails, the flow is supposed to halt. How do we keep throwing the network check failed error multiple times in a flow?

Reply:
Okay, that makes sense